File: TheMindGameJava/src/TheMindGame.java
import ReadModel.Card;
import ReadModel.Player;

import java.util.*;
import java.util.stream.Collectors;

public class TheMindGame implements Runnable {
    public static final int FAILED_GAMES_DEBUG = 2_500_000;
    private final int numPlayers;
    private final int numLevelsToWin;

    private static int failedGames = 0;
    private int currentLevel = 1;
    private static boolean keepRunningThread = true;

    public TheMindGame(int numPlayers, int numLevelsToWin) {
        this.numPlayers = numPlayers;
        this.numLevelsToWin = numLevelsToWin;
    }

    @Override
    public void run() {
        System.out.println("Inside : " + Thread.currentThread().getName());
        TheMindGameResult result = play();
        System.out.println(result);
    }

    public TheMindGameResult play() {
        Map<Integer, List<Card>> successfulPiles = new HashMap<>();

        while (stillLookingForSolution()) {
            List<Card> pileOfCards = new ArrayList<>();
            int desiredPileOfCardsNumber = numPlayers * currentLevel;
            List<Player> players = dealCardsToEachPlayer(numPlayers, currentLevel);

            if (isDebugEnabled()) {
                System.err.println(players);
            }

            do {
                Card currentCard = popRandomlyOnePlayerCard(players);

                if (!isValidCardInPile(currentCard, pileOfCards)) {
                    incrementFailedGames();
                    currentLevel = 1;
                    successfulPiles = new HashMap<>();
                    if (isDebugEnabled()) {
                        System.out.println("Failed game number " + failedGames + ". Trying again... brrr");
                    }
                    break;
                }

                pileOfCards.add(currentCard);
            } while (areStillCardsToPlay(players));

            // Don't level-up if the pile miss cards
            if (pileOfCards.size() < desiredPileOfCardsNumber) {
                continue;
            }

            successfulPiles.put(currentLevel, pileOfCards);
            currentLevel++;
        }

        keepRunningThread = false;

        return new TheMindGameResult(failedGames, successfulPiles);
    }

    private synchronized void incrementFailedGames() {
        failedGames++;
    }

    private boolean stillLookingForSolution() {
        return currentLevel <= numLevelsToWin && keepRunningThread;
    }

    private List<Player> dealCardsToEachPlayer(int numPlayers, int currentLevel) {
        List<Player> players = new ArrayList<>();

        for (int i = 0; i < numPlayers; i++) {
            players.add(Player.create(currentLevel, players));
        }

        return players;
    }

    private boolean isDebugEnabled() {
        return failedGames % FAILED_GAMES_DEBUG == 0;
    }

    private boolean isValidCardInPile(Card currentCard, List<Card> pileOfCards) {
        if (pileOfCards.isEmpty()) {
            return true;
        }

        Card lastCard = pileOfCards.get(pileOfCards.size() - 1);

        return currentCard.number() >= lastCard.number();
    }

    private Card popRandomlyOnePlayerCard(List<Player> players) {
        List<Player> playersWithCards = players
                .stream()
                .filter(Player::hasCards)
                .collect(Collectors.toList());

        int randomPos = new Random().nextInt(playersWithCards.size());
        Player randomPlayer = playersWithCards.get(randomPos);

        return randomPlayer.popMinCard();
    }

    private boolean areStillCardsToPlay(List<Player> players) {
        return players
                .stream()
                .anyMatch(Player::hasCards);
    }
}

File: TheMindGamePhp/src/TheMindGame.php
<?php

declare(strict_types=1);

namespace App;

/**
 * @psalm-immutable
 */
final class TheMindGame
{
    private const FAILED_GAMES_DEBUG = 500_000;
    private static int $failedGames = 0;

    private int $currentLevel = 1;

    public function play(int $numPlayers, int $numLevelsToWin): TheMindGameResult
    {
        $successfulPiles = [];

        while ($this->currentLevel <= $numLevelsToWin) {
            $pileOfCards = [];
            $desiredPileOfCardsNumber = $numPlayers * $this->currentLevel;
            $players = $this->dealCardsToEachPlayer($numPlayers);

            do {
                $currentCard = $this->popRandomlyOnePlayerCard($players);

                if (!$this->isValidCardInPile($currentCard, $pileOfCards)) {
                    static::$failedGames++;
                    $this->currentLevel = 1;
                    $successfulPiles = [];
                    $this->displayDebugInfo($currentCard, $pileOfCards);
                    break;
                }

                $pileOfCards[] = $currentCard;
            } while ($this->areStillCardsToPlay(...$players));

            // Don't level-up if the pile miss cards
            if (count($pileOfCards) < $desiredPileOfCardsNumber) {
                continue;
            }

            $successfulPiles[$this->currentLevel] = $pileOfCards;
            $this->currentLevel++;
        }

        if (count($successfulPiles) !== $numLevelsToWin) {
            return TheMindGameResult::createEmpty();
        }

        return TheMindGameResult::create(static::$failedGames, $successfulPiles);
    }

    /**
     * @return Player[]
     */
    private function dealCardsToEachPlayer(int $numPlayers): array
    {
        $players = [];

        for ($i = 0; $i < $numPlayers; $i++) {
            $players[] = Player::create($this->currentLevel, ...$players);
        }

        return $players;
    }

    /**
     * @param Card[] $pileOfCards
     */
    private function isValidCardInPile(Card $currentCard, array $pileOfCards): bool
    {
        $lastCard = $this->lastPileCard($pileOfCards);
        if (!$lastCard) {
            return true;
        }

        return $currentCard->number() >= $lastCard->number();
    }

    /**
     * @param Card[] $pileOfCards
     */
    private function lastPileCard(array $pileOfCards): ?Card
    {
        $lastCard = end($pileOfCards);

        return (!$lastCard) ? null : $lastCard;
    }

    private function popRandomlyOnePlayerCard(array $players): Card
    {
        /** @var Player[] $playersWithCards */
        $playersWithCards = array_values(array_filter(
            $players,
            static fn(Player $playerCards) => $playerCards->totalCards() > 0
        ));

        $posRandom = random_int(0, count($playersWithCards) - 1);
        $randomPlayer = $playersWithCards[$posRandom];

        return $randomPlayer->popMinCard();
    }

    private function areStillCardsToPlay(Player ...$players): bool
    {
        $playersWithCards = array_filter(
            $players,
            static fn(Player $playerCards) => $playerCards->totalCards() > 0
        );

        return count($playersWithCards) > 0;
    }

    /**
     * @param Card[] $pileOfCards
     */
    private function displayDebugInfo(Card $currentCard, array $pileOfCards)
    {
        if ($this->isDebugEnabled()) {
            echo sprintf(
                "Failed game Nr %d. Trying again... brrr (current:%d, last:%d)\n",
                static::$failedGames,
                $currentCard->number(),
                $this->lastPileCard($pileOfCards)->number()
            );
        }
    }

    private function isDebugEnabled(): bool
    {
        return static::$failedGames % self::FAILED_GAMES_DEBUG === 0;
    }
}
